### ISSUE TITLE
Only 95% accuracy on N-MNIST w/ example 03 (expected ±99.2%)

### ISSUE THREAD

**Issue by Afstkla** [bamsumit/slayerPytorch#58]
I'm trying to achieve the accuracies reported in the SLAYER paper (99.2% on N-MNIST), but I only get to about 95% accuracy on the full (not the NMNISTSmall) 60k train/10k test N-MNIST dataset.

The network is slightly different than the reported topology (example 03 is actually a little bigger than the reported topology if I'm not mistaken), but I'd expect similar results (as the topology seems very similar).

Has anyone been able to run SLAYER and achieve the reported accuracy?
I'm on commit 25c5c6b72a6e271fb3045317e2446860fb40ae6d, with only small changes to make sure I use the full N-MNIST dataset, rather than the small one.

`pip list` returns:
```
Package         Version
--------------- -----------
cycler          0.10.0
future          0.18.2
kiwisolver      1.3.1
matplotlib      2.2.3
numpy           1.18.5
pickleshare     0.7.5
Pillow          8.1.2
pip             21.0.1
pyparsing       2.4.7
python-dateutil 2.8.1
pytz            2021.1
PyYAML          5.4
setuptools      39.2.0
six             1.15.0
slayerCuda      0.0.0
slayerLoihiCuda 0.0.0
slayerSNN       0.0.0
torch           1.0.1.post2
torchvision     0.2.2
tqdm            4.19.9
```

And I'm using CuDNN 7.5.1 and CUDA 10.0.130

Thanks in advance

**Comment by cglagovich** [comment]
What are the details of your training setup? Batch size, learning rate, learning rate scheduler, regularization/weight decay, number of epochs, data augmentation?

**Comment by Afstkla** [comment]
The 95% was achieved with the following configuration:

GPU: 2x 2080 TI (using `DataParallel`)
Batch size: 80 (about the highest batch size that would fit on the GPUs)
Dataset: Full N-MNIST
Other settings (including topology): unchanged from https://github.com/bamsumit/slayerPytorch/blob/master/example/03_NMNIST_CNN/nmnistCNN.py

**Comment by cglagovich** [comment]
I would also like to hear more about the exact training environment to recreate the paper results of 99+% accuracy. I believe that the topology in the example you linked is not the CNN that the SLAYER paper uses to achieve that accuracy, however:

![image](https://user-images.githubusercontent.com/20099419/115039557-8fd42b00-9e9e-11eb-88b1-892fdb4f578e.png)
 Since my interests are putting the model on Loihi, I have been using the Loihi neuron to implement the network. 

```
class PaperCNN(torch.nn.Module):
    '''
    The winning N-MNIST CNN from the SLAYER paper
    '''
    def __init__(self, netParams):
        super(PaperCNN, self).__init__()
        # initialize slayer
        slayer = spikeLayer(netParams['neuron'], netParams['simulation'])
        self.slayer = slayer
        # define network functions
        self.conv1 = slayer.conv(2, 12, 5, padding=1)
        self.conv2 = slayer.conv(12, 64, 5, padding=1)
        self.pool1 = slayer.pool(2)
        self.pool2 = slayer.pool(2)
        self.fc1   = slayer.dense((7*7*64), 10)

    def forward(self, spikeInput):
        x = self.slayer.spikeLoihi(self.conv1(spikeInput)) # 32, 32, 12
        x = self.slayer.delayShift(x, 1)
        x = self.slayer.spikeLoihi(self.pool1(x)) # 16, 16, 12
        x = self.slayer.delayShift(x, 1)
        x = self.slayer.spikeLoihi(self.conv2(x)) # 14, 14, 64
        x = self.slayer.delayShift(x, 1)
        x = self.slayer.spikeLoihi(self.pool2(x)) #  7,  7, 64
        x = self.slayer.delayShift(x, 1)
        x = x.reshape((x.shape[0], -1, 1, 1, x.shape[-1])) # Flatten
        x = self.slayer.spikeLoihi(self.fc1(x)) #  10
        x = self.slayer.delayShift(x, 1)

        return x
```

Maybe you see an issue in the way I implemented it. Training this network with `learning_rate = 0.01` `weight_decay = 0.01` has not given me the results I expect yet.

**Comment by bamsumit** [comment]
The result for the SLAYER paper was from the CUDA/C++ implementation of SLAYER [here](https://bitbucket.org/bamsumit/slayer) so the examples here don't include the exact codes to replicate the results in the paper. Here are the critical parts for PyTorch implementation to get similar/better result. 

* __optimizer__: NADAM works best for me
  `optimizer = snn.utils.optim.Nadam(net.parameters(), lr = 0.01, amsgrad = False)`
* __batch size__: 12 (lower batch size yield better results for me in NMNIST)
* __Spike gradient hyperparameters__
  **`network.yaml`**
  ```
  simulation:
      Ts: 1.0
      tSample: 300
      nSample: 12
  neuron:
      type:     SRMALPHA
      theta:    10
      tauSr:    1.0
      tauRef:   1.0
      scaleRef: 2     # relative to theta
      tauRho:   0.1     # relative to theta 
      scaleRho: 0.25
  training:
      error:
          type: NumSpikes
          tgtSpikeRegion: {start: 0, stop: 300}
          tgtSpikeRate: {true: 60, false: 10}
  ```

* __Data augmentation__
  ```
  def augmentData(event):
      xs = 8
      ys = 8
      xjitter = np.random.randint(2*xs) - xs
      yjitter = np.random.randint(2*ys) - ys
      event.x += xjitter
      event.y += yjitter
      return event
  
  # Dataset definition
  class nmnistDataset(Dataset):
      def __init__(self, datasetPath, sampleFile, samplingTime, sampleLength, augment=False):
          self.path = datasetPath 
          self.samples = np.loadtxt(sampleFile).astype('int')
          self.samplingTime = samplingTime
          self.nTimeBins = int(sampleLength / samplingTime)
          self.augment = augment
  
      def __getitem__(self, index):
          inputIndex  = self.samples[index, 0]
          classLabel  = self.samples[index, 1]
          
          event = snn.io.read2Dspikes(self.path + str(inputIndex.item()) + '.bs2')
          if self.augment is True:    
              event = augmentData(event)
          inputSpikes = event.toSpikeTensor(torch.zeros((2,34,34,self.nTimeBins)),
                          samplingTime=self.samplingTime)
  
          desiredClass = torch.zeros((10, 1, 1, 1))
          desiredClass[classLabel,...] = 1
          return inputSpikes, desiredClass, classLabel
      
      def __len__(self):
          return self.samples.shape[0]
  ```
* __Axonal delays__. It is not critical, but including it will give a slight boost in performance. **Remember to clamp the delays to positive values when using it** 
  ```
  loss.backward()
  optimizer.step()
  module.clamp() # defined in the next bullet
  ```
* __Weight normalization__ for better stability during learning.
  ```
  # Network definition
  class Network(torch.nn.Module):
      def __init__(self, netParams):
          super(Network, self).__init__()
          # initialize slayer
          slayer = snn.layer(netParams['neuron'], netParams['simulation'])
          self.slayer = slayer
          # define network functions
          self.conv1  = torch.nn.utils.weight_norm(slayer.conv( 2, 16, 5, padding=1, weightScale=20), name='weight')
          self.conv2  = torch.nn.utils.weight_norm(slayer.conv(16, 32, 3, padding=1, weightScale=20), name='weight')
          self.conv3  = torch.nn.utils.weight_norm(slayer.conv(32, 64, 3, padding=1, weightScale=20), name='weight')
  
          self.pool1  = slayer.pool(2)
          self.pool2  = slayer.pool(2)
  
          self.fc1    = torch.nn.utils.weight_norm(slayer.dense((8*8*64), 512, weightScale=20), name='weight')
          self.fc2    = torch.nn.utils.weight_norm(slayer.dense(512, 10, weightScale=20), name='weight')
  
          self.delay1 = slayer.delay(16)
          self.delay2 = slayer.delay(16)
          self.delay3 = slayer.delay(32)
          self.delay4 = slayer.delay(32)
          self.delay5 = slayer.delay(64 * 8 * 8)
          self.delay6 = slayer.delay(512)
  
      def forward(self, spike):
          count.append(torch.sum(spike).item())
  
          spike = self.slayer.spike(self.conv1(self.slayer.psp(spike))) # 32, 32, 16
          spike = self.delay1(spike)
          
          spike = self.slayer.spike(self.pool1(self.slayer.psp(spike))) # 16, 16, 16
          spike = self.delay2(spike)
          
          spike = self.slayer.spike(self.conv2(self.slayer.psp(spike))) # 16, 16, 32
          spike = self.delay3(spike)
          
          spike = self.slayer.spike(self.pool2(self.slayer.psp(spike))) #  8,  8, 32
          spike = self.delay4(spike)
          
          spike = self.slayer.spike(self.conv3(self.slayer.psp(spike))) #  8,  8, 64
          spike = spike.reshape((spike.shape[0], -1, 1, 1, spike.shape[-1]))
          spike = self.delay5(spike)
          
          spike = self.slayer.spike(self.fc1  (self.slayer.psp(spike))) #  10
          spike = self.delay6(spike)
          
          spike = self.slayer.spike(self.fc2  (self.slayer.psp(spike))) #  10
          
          return spike
  
      def clamp(self):
          self.delay1.delay.data.clamp_(0, 64)
          self.delay2.delay.data.clamp_(0, 64)
          self.delay3.delay.data.clamp_(0, 64)
          self.delay4.delay.data.clamp_(0, 64)
          self.delay5.delay.data.clamp_(0, 64)
          self.delay6.delay.data.clamp_(0, 64)
  
  
      def gradFlow(self, path):
          gradNorm = lambda x: torch.norm(x).item()/torch.numel(x)
  
          grad = []
          grad.append(gradNorm(self.conv1.weight_g.grad))
          grad.append(gradNorm(self.conv2.weight_g.grad))
          grad.append(gradNorm(self.conv3.weight_g.grad))
          grad.append(gradNorm(self.fc1  .weight_g.grad))
          grad.append(gradNorm(self.fc2  .weight_g.grad))
  
          plt.figure()
          plt.semilogy(grad)
          plt.savefig(path + 'gradFlow.png')
          plt.close()
  ```

**Comment by bamsumit** [comment]
@cglagovich For Loihi, we have better support with SLAYER auto. I do not recommend axonal delay for NMNIST if your end goal is to implement it on the actual chip as it will take excessive resources. Here is the neuron parameter that works with batch size of 128.
```
'neuron' : {
        'type'     : 'LOIHI',
        'vThMant'  : 40,
        'vDecay'   : 128,
        'iDecay'   : 4096,
        'refDelay' : 1,
        'wgtExp'   : 0,
        'tauRho'   : 3,
        'scaleRho' : 256,
    },
```
